File: python/tests/test_cubed_sphere_conformance.py
"""Cross-language conformance for the cubed_sphere grid family.

Verifies that the Python accessor runtime produces accessor outputs that
match the committed Julia-reference golden within the documented
`docs/GRIDS_API.md` §4.2 ULP tolerance. Corpus lives at
``tests/conformance/grids/cubed_sphere/``.
"""

from __future__ import annotations

import json
from pathlib import Path

import pytest

from earthsci_toolkit import grids

HARNESS_DIR = (
    Path(__file__).resolve().parents[2]
    / "tests"
    / "conformance"
    / "grids"
    / "cubed_sphere"
)
FIXTURES_PATH = HARNESS_DIR / "fixtures.json"
GOLDEN_DIR = HARNESS_DIR / "golden"
_SPEC = json.loads(FIXTURES_PATH.read_text())
REL_TOL = float(_SPEC["tolerance"]["relative"])
_FIXTURE_NAMES = [f["name"] for f in _SPEC["fixtures"]]


def _close_rel(a: float, b: float, tol: float = REL_TOL) -> bool:
    scale = max(1.0, abs(a), abs(b))
    return abs(a - b) <= tol * scale


def _fixture(name: str) -> dict:
    for f in _SPEC["fixtures"]:
        if f["name"] == name:
            return f
    raise KeyError(name)


@pytest.mark.parametrize("fixture_name", _FIXTURE_NAMES)
def test_conformance_against_golden(fixture_name: str) -> None:
    fixture = _fixture(fixture_name)
    opts = fixture["opts"]
    grid = grids.cubed_sphere(
        Nc=opts["Nc"], R=opts["R"], ghosts=opts["ghosts"], dtype=opts["dtype"]
    )
    golden = json.loads((GOLDEN_DIR / f"{fixture_name}.json").read_text())

    assert grid.n_cells == golden["n_cells"]
    assert len(fixture["query_points"]) == len(golden["cell_centers"])

    metric_names = (
        "J",
        "g_xixi",
        "g_etaeta",
        "g_xieta",
        "ginv_xixi",
        "ginv_etaeta",
        "ginv_xieta",
    )

    for k, (p, i, j) in enumerate(fixture["query_points"]):
        lon, lat = grid.cell_centers(p=p, i=i, j=j)
        gc = golden["cell_centers"][k]
        assert _close_rel(lon, gc["lon"]), f"lon mismatch at qp[{k}]=({p},{i},{j})"
        assert _close_rel(lat, gc["lat"]), f"lat mismatch at qp[{k}]=({p},{i},{j})"

        nbrs = grid.neighbors(p, i, j)
        for dkey in ("W", "E", "S", "N"):
            got = list(nbrs[dkey])
            expected = [int(x) for x in golden[f"neighbors_{dkey}"][k]]
            assert got == expected, f"neighbor {dkey} mismatch at qp[{k}]=({p},{i},{j})"

        for name in metric_names:
            got = grid.metric_eval(name, p, i, j)
            expected = float(golden[f"metric_{name}"][k])
            assert _close_rel(got, expected), (
                f"metric {name} mismatch at qp[{k}]=({p},{i},{j}): got {got}, expected {expected}"
            )

        got_area = grid.metric_eval("area", p, i, j)
        expected_area = float(golden["area"][k])
        assert _close_rel(got_area, expected_area), (
            f"area mismatch at qp[{k}]=({p},{i},{j})"
        )

File: test/test_cubed_sphere_conformance.jl
@testsnippet CubedSphereConformanceSetup begin
    using Test
    using EarthSciDiscretizations
    using JSON

    const _CS_HARNESS_DIR = joinpath(@__DIR__, "..", "tests", "conformance", "grids", "cubed_sphere")
    const _CS_REPO_ROOT = joinpath(_CS_HARNESS_DIR, "..", "..", "..", "..")
end

# Mirrors the 0-indexed panel-connectivity logic committed to
# tests/conformance/grids/cubed_sphere/regenerate_golden.jl. Kept inline here
# because Julia's CubedSphereGrid does not yet expose a first-class
# `neighbors(p, i, j)` accessor; the cross-language conformance contract is
# verified by computing neighbors from the same `PANEL_CONNECTIVITY` table.
@testitem "Cubed-sphere cross-language conformance" setup = [CubedSphereConformanceSetup] tags = [:conformance, :grid] begin
    HARNESS_DIR = joinpath(@__DIR__, "..", "tests", "conformance", "grids", "cubed_sphere")
    FIXTURES = JSON.parsefile(joinpath(HARNESS_DIR, "fixtures.json"))
    REL_TOL = Float64(FIXTURES["tolerance"]["relative"])

    DIRECTIONS = (West, East, South, North)
    DIR_KEYS = ("W", "E", "S", "N")

    function neighbor_0idx(p::Int, i::Int, j::Int, dir, Nc::Int)
        if dir == West && i > 0
            return (p, i - 1, j)
        elseif dir == East && i < Nc - 1
            return (p, i + 1, j)
        elseif dir == South && j > 0
            return (p, i, j - 1)
        elseif dir == North && j < Nc - 1
            return (p, i, j + 1)
        end
        nb = PANEL_CONNECTIVITY[p + 1][dir]
        nb_panel = nb.neighbor_panel - 1
        nb_edge = nb.neighbor_edge
        along = (dir == West || dir == East) ? j : i
        if nb.reverse_index
            along = Nc - 1 - along
        end
        if nb_edge == West
            return (nb_panel, 0, along)
        elseif nb_edge == East
            return (nb_panel, Nc - 1, along)
        elseif nb_edge == South
            return (nb_panel, along, 0)
        else
            return (nb_panel, along, Nc - 1)
        end
    end

    function close_rel(a::Real, b::Real, tol::Real)
        scale = max(1.0, abs(a), abs(b))
        return abs(a - b) <= tol * scale
    end

    for fixture in FIXTURES["fixtures"]
        name = fixture["name"]
        opts = fixture["opts"]
        Nc = Int(opts["Nc"])
        R = float(opts["R"])
        ghosts = Int(opts["ghosts"])
        golden = JSON.parsefile(joinpath(HARNESS_DIR, "golden", "$name.json"))

        @testset "fixture=$name" begin
            grid = CubedSphereGrid(Nc; R = R, Ng = ghosts)
            @test 6 * Nc * Nc == golden["n_cells"]

            for (k, qp) in enumerate(fixture["query_points"])
                p, i, j = Int(qp[1]), Int(qp[2]), Int(qp[3])

                lon, lat = gnomonic_to_lonlat(grid.ξ_centers[i + 1], grid.η_centers[j + 1], p + 1)
                gc = golden["cell_centers"][k]
                @test close_rel(lon, gc["lon"], REL_TOL)
                @test close_rel(lat, gc["lat"], REL_TOL)

                for (dkey, dir) in zip(DIR_KEYS, DIRECTIONS)
                    np, ni, nj = neighbor_0idx(p, i, j, dir, Nc)
                    gn = golden["neighbors_$dkey"][k]
                    @test [np, ni, nj] == [Int(gn[1]), Int(gn[2]), Int(gn[3])]
                end

                J, gxx, gyy, gxy = gnomonic_metric(
                    grid.ξ_centers[i + 1], grid.η_centers[j + 1], R,
                )
                det = gxx * gyy - gxy * gxy
                @test close_rel(J, golden["metric_J"][k], REL_TOL)
                @test close_rel(gxx, golden["metric_g_xixi"][k], REL_TOL)
                @test close_rel(gyy, golden["metric_g_etaeta"][k], REL_TOL)
                @test close_rel(gxy, golden["metric_g_xieta"][k], REL_TOL)
                @test close_rel(gyy / det, golden["metric_ginv_xixi"][k], REL_TOL)
                @test close_rel(gxx / det, golden["metric_ginv_etaeta"][k], REL_TOL)
                @test close_rel(-gxy / det, golden["metric_ginv_xieta"][k], REL_TOL)

                area = compute_cell_area(
                    (grid.ξ_edges[i + 1], grid.ξ_edges[i + 2]),
                    (grid.η_edges[j + 1], grid.η_edges[j + 2]),
                    R, p + 1,
                )
                @test close_rel(area, golden["area"][k], REL_TOL)
            end

            # Rule-eval conformance: verify that eval_coeff (canonical ESS passthrough)
            # reproduces the per-axis stencil coefficients committed in the golden.
            # bindings_per_qp are h (uniform grid spacing) and J (Jacobian at cell center).
            # eval_coeff must match the golden stencil_coeffs to REL_TOL.
            rule_blocks = get(golden, "rule_evals", [])
            @test !isempty(rule_blocks)  # golden missing rule_evals → regenerate with regenerate_golden.jl
            for block in rule_blocks
                rule_name = block["rule"]
                rule_path = joinpath(_CS_REPO_ROOT, block["rule_path"])
                payload = JSON.parsefile(rule_path)
                rule_def = payload["discretizations"][rule_name]
                stencil = rule_def["stencil"]
                golden_coeffs = block["stencil_coeffs"]
                golden_bindings = block["bindings_per_qp"]

                @testset "rule_eval/$rule_name" begin
                    @test length(golden_coeffs) == length(fixture["query_points"])
                    for (k, qp) in enumerate(fixture["query_points"])
                        p, i, j = Int(qp[1]), Int(qp[2]), Int(qp[3])
                        h = π / (2 * Nc)
                        J_val, = gnomonic_metric(grid.ξ_centers[i + 1], grid.η_centers[j + 1], R)
                        binds = Dict{String, Float64}("h" => h, "J" => J_val)

                        # Binding check: golden bindings_per_qp must match the
                        # values extracted from the Julia grid accessor.
                        for (bname, bval) in golden_bindings[k]
                            got = binds[bname]
                            expected = Float64(bval)
                            @test close_rel(got, expected, REL_TOL)
                        end

                        # Coefficient check: eval_coeff on the rule's AST must match
                        # the golden stencil_coeffs row.
                        golden_row = golden_coeffs[k]
                        @test length(stencil) == length(golden_row)
                        for (s_idx, entry) in enumerate(stencil)
                            got = eval_coeff(entry["coeff"], binds)
                            expected = Float64(golden_row[s_idx])
                            @test close_rel(got, expected, REL_TOL)
                        end
                    end
                end
            end
        end
    end
end
